wrapanapi/systems/rhevm.py

```diff
@@ -1148,31 +1148,94 @@ def _get_attached_storage_domain_service(self, datacenter_id, storage_domain_id)
+        desired_state = getattr(types.StorageDomainStatus, state.upper(), None)
+        active = types.StorageDomainStatus.ACTIVE
+        if desired_state is None:
+            raise ValueError('Invalid state [{}] passed for setting storage domain, '
+                             'value values are {}'.format(state, list(types.StorageDomainStatus)))
+        for datacenter in self._data_centers_service.list():
+            for domain in self.api.follow_link(datacenter.storage_domains):
                 if domain.name == storage_domain_name:
-                    asds = self._get_attached_storage_domain_service(dc.id, domain.id)
-                    if state == "maintenance" and domain.status.value == "active":
-                        asds.deactivate()
-                    elif state == "active" and domain.status.value != "active":
-                        asds.activate()
-                    wait_for(lambda: domain.status.value == state, delay=5, num_sec=240)
+                    attached_service = self._get_attached_storage_domain_service(datacenter.id,
+                                                                                 domain.id)
+                    domain_status = self.api.follow_link(domain).status
+                    if domain_status == desired_state:
+                        return None  # already on the state we wanted
+                    elif desired_state != active:
+                        attached_service.deactivate()
+                        expected_state = types.StorageDomainStatus.MAINTENANCE
+                    else:
+                        attached_service.activate()
+                        expected_state = active
+                    wait_for(
+                        lambda: self.api.follow_link(domain).status == expected_state,
+                        delay=5,
+                        num_sec=timeout
+                    )
                     return True
-        return False
+        else:
+            # domain name was never matched on any data center
+            raise ValueError('Given domain name [{}] was never matched'.format(storage_domain_name))
 
-    def get_template_from_storage_domain(self, template_name, storage_domain_name):
+    def get_template_from_storage_domain(
+        self, template_name, storage_domain_name, unregistered=False
+    ):
+        """get a specific named template on a given storage domain
+
+        Args:
+            template_name (str): name of the template to get
+            storage_domain_name (str): name of the storage domain to get a template on
+            unregistered (bool): passed to ovirt TemplatesService.list()
+
+        Raises:
+            exceptions that ovirt returns, ItemNotFound if the storage_domain_name is bad
+        """
         sds = self._get_storage_domain_service(storage_domain_name)
-        for template in sds.templates_service().list(unregistered=False):
+        for template in sds.templates_service().list(unregistered=unregistered):
```

Comment:
What does the unregistered do? As I see it, if its False, it will list registered templates but what when its True?

Reply:
http://ovirt.github.io/ovirt-engine-sdk/master/services.m.html#ovirtsdk4.services.StorageDomainTemplatesService.list

```
unregistered
Indicates whether to retrieve a list of registered or unregistered templates which contain disks on the storage domain. To get a list of unregistered templates the call should indicate the unregistered flag. For example, to get a list of unregistered templates the REST API call should look like this:

GET /ovirt-engine/api/storagedomains/123/templates?unregistered=true

The default value of the unregisterd flag is false. The request only apply to storage domains that are attached.
```

Reply:
Just turning it into an optional kwarg, maintaining the same behavior as before.

Reply:
Thanks for explanation. Whole concept of unregistered template is new to me. Seems counter-intuitive. Maybe it means templates those are on storage domain but not registered, essentially available for import. Is it?

Reply:
I'm not exactly sure what they're referencing either, they don't detail in the API docs what that specifically means, and a cursory search in the domain documentation didn't bring anything up either.